### PR TITLE
List.fold/contains simplification

### PR DESCRIPTION
**Pull request by ncave**
Generates simpler IL.

### CODE REVIEW

File: src/fsharp/FSharp.Core/list.fs [hunk 1]
```diff
@@ -356,12 +352,10 @@ namespace Microsoft.FSharp.Collections
         let exists predicate list = Microsoft.FSharp.Primitives.Basics.List.exists predicate list
         
         [<CompiledName("Contains")>]
-        let inline contains value source =
-            let rec contains e xs1 =
-                match xs1 with
-                | [] -> false
-                | h1::t1 -> e = h1 || contains e t1
-            contains value source
+        let rec contains value source =
```
Comment (isaacabraham) [inline src/fsharp/FSharp.Core/list.fs]: Does it matter that contains is no longer marked as `inline`?

Comment (ncave) [reply src/fsharp/FSharp.Core/list.fs]: @isaacabraham idk, but neither are `exists`, `find/findBack`, `tryFind/tryFindBack`, `pick/tryPick` etc.

Comment (isaacabraham) [reply src/fsharp/FSharp.Core/list.fs]: Hmmm - so why was `contains` inlined? Just some historical artifact, or some reason.

Comment (forki) [reply src/fsharp/FSharp.Core/list.fs]: People started to inline stuff in other PRs...

**Comment by KevinRansom** [comment]
Thank you for this.

**Comment by dsyme** [comment]
@KevinRansom Please revert this, it needs proper review.   And can we in general please look at changes a bit more deeply before integrating them?

For example, it removes an optimization for empty lists here.https://github.com/Microsoft/visualfsharp/pull/3975/files#diff-9c7a3b0c1ddcc1a72b924ffac75f21feL215.  Perhaps we can do this, but it needs performance testing and review.

It also removes the use of ``inline`` but I haven't looked at that yet - the first issue is enough to mean that it shouldn't be integrated.

File: src/fsharp/FSharp.Core/list.fs (hunk 1 above)
Comment (dsyme) [reply src/fsharp/FSharp.Core/list.fs]: Yes, the "inline" matters a lot.  The inlining causes the equality check to become monomorphic (non-generic) in the common case that the callsite is monmorphic, and hence the equality check will be type-specialized

File: src/fsharp/FSharp.Core/list.fs [hunk 2]
```diff
@@ -212,15 +212,11 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Fold")>]
         let fold<'T,'State> folder (state:'State) (list: 'T list) = 
-            match list with 
-            | [] -> state
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/list.fs]: This optimization for the case where the list is empty has been removed - can it please be added back in?  Or lets just not do this change

**Comment by vasily-kirichenko** [comment]
@dsyme the problem is that we never know when (and if) you are going to review a PR. Is it possible to mark a PR as "must be reviewed by ....", with proper notification? (or "I'm gonna review this by the end of this week")

File: src/fsharp/FSharp.Core/list.fs (hunk 2 above)
Comment (ncave) [reply src/fsharp/FSharp.Core/list.fs]: @dsyme Thanks, I'll add it. The optimization for the empty case is still there to some extent, in the sense that the folding function is not being called in the "empty list" case. But yes, additional check in the beginning for the empty case makes a lot of sense, so the Adapt function is not called prematurely. I'll resubmit.

**Comment by dsyme** [comment]
* Let's all strive for high standard reviews
* I try to review everything except Visual F# IDE Tooling specific things which I just glance at.  Assign me as a reviewer and @dsyme me if needed - after a weekend I tend to search for those first
* PRs should come with explanations, thanks

File: src/fsharp/FSharp.Core/list.fs (hunk 2 above)
Comment (dsyme) [reply src/fsharp/FSharp.Core/list.fs]: Thanks. I integrated the revert so you can submit a new PR at your leisure :) thanks all!

File: src/fsharp/FSharp.Core/list.fs (hunk 1 above)
Comment (ncave) [reply src/fsharp/FSharp.Core/list.fs]: @dsyme I'll take it out if the inline is that important, as I'm not sure how to get around FS1118 (cannot inline recursive).
There are a few other funcs with `when 'T : equality` signature (except, distinct, countBy etc.), should they be inline as well?

Comment (dsyme) [reply src/fsharp/FSharp.Core/list.fs]: Yes, the nested ``let rec`` is to get around the restriction on ``let rec inline``.  The whole inner ``let rec`` gets copied that way.

Was the use of a nested ``let rec`` causing problems for Fable?  If so you may often be able to replace with an explicit loop

Comment (ncave) [reply src/fsharp/FSharp.Core/list.fs]: @dsyme Nothing to do with Fable, this optimization was just to remove the unnecessary extra object allocation and call that is generated, but I understand the inline modifier is more important here (although there are other funcs with `when 'T : equality` that are not inline).

Comment (dsyme) [reply src/fsharp/FSharp.Core/list.fs]: > I understand the inline modifier is more important here (although there are other funcs with when 'T : equality that are not inline).

If there is  performance evidence to show that it's worthwhile inlining these then we should likely do it.  Normally it's easy to get perf evidence in such cases. Sometimes the inlining can't be done due to the implementation depending on other internal functions we don't want to make public

**Pull request merged**